### PR TITLE
Handle error cases gracefully. WASM -> Wasm.

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

### CODE REVIEW

File: include/proxy-wasm/wasm.h [hunk 1]
```diff
@@ -289,6 +291,9 @@ inline const std::string &WasmBase::vm_configuration() const {
 }
 
 inline void *WasmBase::allocMemory(uint64_t size, uint64_t *address) {
+  if (!malloc_) {
+    return nullptr;
+  }
```
Comment (PiotrSikora) [inline include/proxy-wasm/wasm.h:296]: Do we need this? We fail hard if `malloc` is not exported.

(I'm not saying it's not good to add it just to be safe, I'm just curious why did you add it now)

File: src/wasm.cc [hunk 2]
```diff
@@ -508,7 +533,7 @@ getOrCreateThreadLocalWasm(std::shared_ptr<WasmHandleBase> base_wasm,
   if (wasm_handle) {
     auto root_context = wasm_handle->wasm()->getOrCreateRootContext(plugin);
     if (!wasm_handle->wasm()->configure(root_context, plugin)) {
-      base_wasm->wasm()->error("Failed to configure WASM code");
+      base_wasm->wasm()->error("Failed to configure Wasm code");
```
Comment (PiotrSikora) [inline src/wasm.cc:536]: `s/Wasm code/thread-local Wasm plugin/`

File: include/proxy-wasm/wasm.h (hunk 1 above)
Comment (jplevyak) [reply include/proxy-wasm/wasm.h:296]: Generally I was trying to make the proxy-independent code resilient against unexpected behavior.  An integrator could choose to not hard fail on error().  For envoy we throw an exception which we will catch at a higher level, but presumably someone code set a flag.

File: src/wasm.cc [hunk 3]
```diff
@@ -445,44 +443,71 @@ std::shared_ptr<WasmHandleBase> createWasm(std::string vm_key, std::string code,
...
+      wasm_handle = it->second.lock();
+      if (!wasm_handle) {
         base_wasms->erase(it);
       }
     }
-    if (wasm)
-      return wasm;
-    wasm = factory(vm_key);
-    (*base_wasms)[vm_key] = wasm;
+    if (wasm_handle)
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply src/wasm.cc]: Done

File: src/wasm.cc (hunk 2 above)
Comment (jplevyak) [reply src/wasm.cc]: Done.
